File: packages/sdk/src/administrator/admin-lambdas.ts
import BaseSDK from '../base-sdk';
import { QlError } from '../types';

export type ManageLambdasSDK = {
  getList: (sourceId: string) => Promise<any>;
  getLambda: (sourceId: string, functionName: string) => Promise<any>;
  create: (sourceId: string, data: any) => Promise<any>;
  update: (sourceId: string, functionName: string, data: any) => Promise<any>;
  delete: (sourceId: string, functionName: string) => Promise<any>;
};

export default (sdk: BaseSDK): { manageLambdas: ManageLambdasSDK } => ({
  manageLambdas: {
    getList: (sourceId: string) => {
      if (!sourceId) {
        throw new QlError('Source ID is required');
      }
      return sdk.get(`/lambdas/${sourceId}`);
    },
    getLambda: (sourceId: string, functionName: string) => {
      if (!sourceId || !functionName) {
        throw new QlError('Source ID and function name are required');
      }
      return sdk.get(`/lambdas/${sourceId}/${functionName}`);
    },
    create: (sourceId: string, data: any) => {
      if (!sourceId) {
        throw new QlError('Source ID is required');
      }
      return sdk.post(`/lambdas/${sourceId}`, data);
    },
    update: (sourceId: string, functionName: string, data: any) => {
      if (!sourceId || !functionName) {
        throw new QlError('Source ID and function name are required');
      }
      return sdk.put(`/lambdas/${sourceId}/${functionName}`, data);
    },
    delete: (sourceId: string, functionName: string) => {
      if (!sourceId || !functionName) {
        throw new QlError('Source ID and function name are required');
      }
      return sdk.delete(`/lambdas/${sourceId}/${functionName}`);
    },
  },
});

File: packages/python-sdk/qelos_sdk/administrator/admin_lambdas.py
from __future__ import annotations

import json
from typing import Any, Dict, List

from ..base_sdk import BaseSDK
from ..types import QelosSDKOptions


class QlManageLambdas(BaseSDK):
    """Admin lambda/serverless function management."""

    _relative_path = "/api/lambdas"

    def __init__(self, options: QelosSDKOptions) -> None:
        super().__init__(options)

    async def get_list(self, source_id: str) -> List[Any]:
        """List lambdas for a source."""
        return await self.call_json_api(f"{self._relative_path}/{source_id}")

    async def get_lambda(self, source_id: str, function_name: str) -> Any:
        """Get a specific lambda."""
        return await self.call_json_api(f"{self._relative_path}/{source_id}/{function_name}")

    async def create(self, source_id: str, data: Any) -> Any:
        """Create a new lambda."""
        return await self.call_json_api(
            f"{self._relative_path}/{source_id}",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(data),
        )

    async def update(self, source_id: str, function_name: str, data: Any) -> Any:
        """Update a lambda."""
        return await self.call_json_api(
            f"{self._relative_path}/{source_id}/{function_name}",
            method="PUT",
            headers={"content-type": "application/json"},
            body=json.dumps(data),
        )

    async def remove(self, source_id: str, function_name: str) -> Any:
        """Delete a lambda."""
        return await self.call_api(
            f"{self._relative_path}/{source_id}/{function_name}", method="DELETE"
        )
